compiler/packages/babel-plugin-react-compiler/src/Transform/TransformFire.ts

```diff
@@ -0,0 +1,633 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import {CompilerError, CompilerErrorDetailOptions, ErrorSeverity} from '..';
+import {
+  CallExpression,
+  Effect,
+  Environment,
+  FunctionExpression,
+  GeneratedSource,
+  HIRFunction,
+  Identifier,
+  IdentifierId,
+  Instruction,
+  InstructionId,
+  InstructionKind,
+  InstructionValue,
+  isUseEffectHookType,
+  LoadLocal,
+  makeInstructionId,
+  Place,
+  promoteTemporary,
+} from '../HIR';
+import {createTemporaryPlace, markInstructionIds} from '../HIR/HIRBuilder';
+import {getOrInsertWith} from '../Utils/utils';
+import {BuiltInFireId, DefaultNonmutatingHook} from '../HIR/ObjectShape';
+
+/*
+ * TODO(jmbrown):
+ *   In this stack:
+ *     - Insert useFire import
+ *     - Assert no lingering fire calls
+ *     - Ensure a fired function is not called regularly elsewhere in the same effect
+ *
+ *   Future:
+ *   - rewrite dep arrays
+ *   - traverse object methods
+ *   - method calls
+ */
+
+const CANNOT_COMPILE_FIRE = 'Cannot compile `fire`';
+
+function deleteInstructions(
+  deleteInstrs: Set<InstructionId>,
+  instructions: Array<Instruction>,
+): Array<Instruction> {
+  if (deleteInstrs.size > 0) {
+    const newInstrs = instructions.filter(instr => !deleteInstrs.has(instr.id));
+    return newInstrs;
+  }
+  return instructions;
+}
+
+function rewriteInstructions(
+  rewriteInstrs: Map<InstructionId, Array<Instruction>>,
+  instructions: Array<Instruction>,
+): Array<Instruction> {
+  if (rewriteInstrs.size > 0) {
+    const newInstrs = [];
+    for (const instr of instructions) {
+      const newInstrsAtId = rewriteInstrs.get(instr.id);
+      if (newInstrsAtId != null) {
+        newInstrs.push(...newInstrsAtId, instr);
+      } else {
+        newInstrs.push(instr);
+      }
+    }
+
+    return newInstrs;
+  }
+
+  return instructions;
+}
+
+function makeLoadUseFireInstruction(env: Environment): {
+  loadUseFireInstr: Instruction;
+  useFirePlace: Place;
+} {
+  const useFirePlace = createTemporaryPlace(env, GeneratedSource);
+  useFirePlace.effect = Effect.Read;
+  useFirePlace.identifier.type = DefaultNonmutatingHook;
```

Comment:
can we be more precise and use the BuiltInFireId type?

Reply:
yeah fair. just thinking about being precise, it would honestly only really help with debugging since the printed HIR would have the type